safir/src/safir/schema_manager/_pydantic_schema_manager.py

```diff
@@ -0,0 +1,229 @@
+        #
+        # In these cases, we want to raise an error if the candidate schema is
+        # incompatible. The only way to accomplish this is to explicitly check
+        # compatibility with the latest version in the registry.
+        #
+        # NOTE: This also doesn't update the 'latest' version to the candidate
+        # schema after a register call of the candidate schema. "latest"
+        # remains at the last registered non-repeat version.
+        schema = model.avro_schema_to_python()
+        result = await self._registry.test_compatibility(
+            subject=subject, schema=schema, verbose=True
+        )
+        if not isinstance(result, dict):
+            raise TypeError(
+                "There seems to be a bug in the python-schema-registry-client"
+                " library!"
+            )
+        if result["is_compatible"] is False:
+            raise IncompatibleSchemaError(result["messages"])
+
+        schema_id = await self._registry.register(subject, schema)
+        self._subjects_to_ids[subject] = schema_id
+
+        return SchemaInfo(schema=schema, schema_id=schema_id, subject=subject)
+
+    async def serialize(self, data: AvroBaseModel) -> bytes:
+        """Serialize the data.
+
+        The model's schema must have been registered before calling this
+        method.
+
+        Parameters
+        ----------
+        data
+            The data to serialize.
+
+        Returns
+        -------
+        bytes
+            The serialized data in the Confluent Wire Format.
+        """
+        subject = self._get_subject(data)
+        try:
+            schema_id = self._subjects_to_ids[subject]
+        except KeyError:
+            raise UnknownSchemaError(
+                f"Schema for model: {data} with subject: {subject} is not"
+                " known to the manager. ``register`` must be called before you"
+                " try to serialize instances of this model."
+            ) from None
+        return await self._serializer.encode_record_with_schema_id(
+            schema_id, data.model_dump()
+        )
+
+    def _get_subject(self, model: AvroBaseModel | type[AvroBaseModel]) -> str:
+        """Get a unique subject for this model based on metadata and config."""
+        # Mypy can't detect the Meta class on the model, so we have to ignore
+        # those lines.
+
+        klass = model if inspect.isclass(model) else model.__class__
+
+        try:
+            name = klass.Meta.schema_name  # type: ignore [union-attr]
+            if not isinstance(name, str):
+                raise InvalidMetadataError("Meta.schema_name must be a str")
+        except AttributeError:
+            name = klass.__name__
+
+        try:
+            namespace = klass.Meta.namespace  # type: ignore [union-attr]
+            if not isinstance(namespace, str):
+                raise InvalidMetadataError("Meta.namespace must be a str")
+        except AttributeError:
+            namespace = None
```

Comment:
The typing here is pretty ugly. I feel like there should be some better way to do this, but I'd have to do some research to see if there's some way to properly type this. The nested subclass thing is kind of a nightmare for typing, which is why Pydantic moved away from it in v2. Some judicious use of `getattr` might help here?

I hate to have run-time type checks when the whole point of mypy is to do that during development and not have to write explicit code to do them, but there may just not be a way around that if mypy can't understand that certain classes have embedded classes that implement a particular protocol (other than blindly stringifying whatever we get, which honestly wouldn't be that unreasonable).

Reply:
I did have a way to typecheck this... I don't remember exactly what I did, but it was pretty nasty, I think it did do a bunch of `getattr`. I eventually decided to leave this as is, because getting it to a place where I could get rid of the `# type: ignore` didn't make anything any safer and made the code much more confusing.

I'd be glad to dig it up again, but I'd like one more assurance that you really want to see it :)

Reply:
...and I missed the whole point of your comment and you're talking about the `isinstance` checks of the `Meta` attributes.  Yeah, I'm not sure if there's a better way to do that, but I like the idea of just stringifying whatever they contain.  I'll do that if I can't find some fancy way of making that check static.

Reply:
Yeah, I ended up back at a bunch of nested `getattr`s, so I just `str`'d the attribute and got rid of the isinstance checks: https://github.com/lsst-sqre/safir/pull/303/commits/0faa5fd4a6e42e0e2e631261bbc64dbe3e5037dc

If you like nested getattrs better, I'll change it.

Reply:
Yeah, that seems fine. I think that's as good as we can do. https://github.com/python/mypy/issues/15835 is the feature that I really want here, but it appears to be a hard problem and no one has found a good solution yet.